File: app/src/main/java/org/mifos/mobile/ui/activities/TwoFactorAuthActivity.java
package org.mifos.mobile.ui.activities;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.Toast;

import com.warrenstrange.googleauth.GoogleAuthenticator;

import org.mifos.mobile.R;
import org.mifos.mobile.ui.activities.base.BaseActivity;
import org.mifos.mobile.utils.Constants;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author Prashant Khandelwal
 * On 11/04/19.
 */
public class TwoFactorAuthActivity extends BaseActivity {

    @BindView(R.id.et_code)
    EditText etCode;

    private GoogleAuthenticator gAuth;
    private String sharedKey;
    private SharedPreferences sharedPreferences;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_two_factor_auth);
        showBackButton();

        ButterKnife.bind(this);
        setToolbarTitle(getString(R.string.title_verify_code));

        sharedPreferences = getSharedPreferences(Constants.TWO_FACTOR_AUTHENTICATION,
                Context.MODE_PRIVATE);
        sharedKey = sharedPreferences.getString(Constants.TWO_FACTOR_AUTHENTICATION_KEY, "");
        System.setProperty("com.warrenstrange.googleauth.rng.algorithmProvider", "AndroidOpenSSL");
        gAuth = new GoogleAuthenticator();

    }

    @OnClick(R.id.btn_submit)
    void verify() {
        String stCode = etCode.getText().toString();
        if (!stCode.isEmpty()) {
            try {
                Integer code = Integer.parseInt(stCode);
                boolean isCodeValid = gAuth.authorize(sharedKey, code);
                if (isCodeValid) {
                    startPassCodeActivity();
                } else {
                    Toast.makeText(this, getString(R.string.incorrect_code),
                            Toast.LENGTH_SHORT).show();
                }
            } catch (Exception e) {
                Toast.makeText(this, getString(R.string.invalid_code),
                        Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(this, getString(R.string.enter_verification_code),
                    Toast.LENGTH_SHORT).show();
        }
    }

    /**
     * Starts {@link PassCodeActivity} with {@code Constans.INTIAL_LOGIN} as true
     */
    private void startPassCodeActivity() {
        Intent intent = new Intent(this, PassCodeActivity.class);
        intent.putExtra(Constants.INTIAL_LOGIN, true);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/activities/TwoFactorAuthActivity.kt
package org.mifos.mobile.ui.activities

import android.app.Activity
import android.os.Bundle
import androidx.lifecycle.ViewModelProviders
import kotlinx.android.synthetic.main.activity_two_factor_auth.*
import org.mifos.mobile.R
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.viewModels.TwoFactorAuthViewModel
import org.mifos.mobile.viewModels.TwoFactorAuthViewModelFactory
import javax.inject.Inject

/**
 * @author Prashant Khandelwal
 * On 11/04/19.
 */
class TwoFactorAuthActivity : BaseActivity() {

    @Inject
    lateinit var twoFactorAuthViewModelFactory: TwoFactorAuthViewModelFactory

    private var userName: String? = null

    private lateinit var viewModel: TwoFactorAuthViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_two_factor_auth)
        showBackButton()

        setToolbarTitle(getString(R.string.title_verify_code))
        activityComponent.inject(this)

        viewModel = ViewModelProviders.of(this, twoFactorAuthViewModelFactory)
                .get(TwoFactorAuthViewModel::class.java)

        if (intent != null && intent.hasExtra(Constants.USER_NAME)) {
            userName = intent.getStringExtra(Constants.USER_NAME)
        }

        btn_submit.setOnClickListener {
            verify()
        }
    }

    private fun verify() {
        val stCode = et_code.text.toString()
        if (stCode.isNotEmpty()) {
            try {
                val code = Integer.parseInt(stCode)
                val isCodeValid = viewModel.verifyCode(code)
                if (isCodeValid) {
                    intent.putExtra(Constants.USER_NAME, userName)
                    setResult(Activity.RESULT_OK, intent)
                    finish()
                } else {
                    showToast(getString(R.string.incorrect_code))
                }
            } catch (e: Exception) {
                showToast(getString(R.string.invalid_code))
            }

        } else {
            showToast(getString(R.string.enter_verification_code))
        }
    }
}
